### PR TITLE
FEM-2489 - Artwork view for audio only media content

### PR DESCRIPTION
**Pull request by GouravSna**
- Artwork view for audio only media content.

### CODE REVIEW

File: playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerView.java [hunk 1]
```diff
@@ -18,24 +18,27 @@
...
 import android.widget.FrameLayout;
+import android.widget.ImageView;
 
 import com.google.android.exoplayer2.Player;
 import com.google.android.exoplayer2.SimpleExoPlayer;
 import com.google.android.exoplayer2.text.Cue;
 import com.google.android.exoplayer2.text.TextOutput;
 import com.google.android.exoplayer2.ui.AspectRatioFrameLayout;
 import com.google.android.exoplayer2.ui.SubtitleView;
+import com.google.android.exoplayer2.ui.spherical.SphericalSurfaceView;
```
Comment (giladna) [inline playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerView.java:39]: @GouravSna  why is this needed?

File: playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerView.java [hunk 2]
```diff
@@ -250,7 +254,17 @@ public void showVideoSubtitles() {
     public SubtitleView getSubtitleView() {
         return subtitleView;
     }
-    
+
+    @Override
+    public void artworkViewVisibility(boolean visibility) {
```
Comment (giladna) [inline playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerView.java]: @GouravSna  please rename to setArtwoekViewVisibility

File: playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerView.java [hunk 3]
```diff
@@ -417,5 +439,32 @@ public void setSurfaceAspectRatioResizeMode(PKAspectRatioResizeMode resizeMode)
...
         return exoPlayerAspectRatioResizeMode;
     }
+
+    private void setArtworkDrawableView(@Nullable Drawable drawable) {
+        if (drawable != null) {
+            int width  = drawable.getIntrinsicWidth();
+            int height = drawable.getIntrinsicHeight();
+            if (width > 0 && height > 0) {
+                float artworkAspectRatio = (float) width / height;
+                onContentAspectRatioChanged(artworkAspectRatio, contentFrame, artworkView);
```
Comment (giladna) [inline playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerView.java]: why is this needed?

File: playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerView.java [hunk 4]
```diff
@@ -417,5 +439,32 @@ public void setSurfaceAspectRatioResizeMode(PKAspectRatioResizeMode resizeMode)
...
+        } else {
+            artworkViewVisibility(false);
+            log.e("Passed drawable for artwork view is null.");
+        }
+    }
+
+    private void onContentAspectRatioChanged(float contentAspectRatio, @Nullable AspectRatioFrameLayout contentFrame, @Nullable View contentView) {
+        if (contentFrame != null) {
+            contentFrame.setAspectRatio(
+                    contentView instanceof SphericalSurfaceView ? 0 : contentAspectRatio);
```
Comment (giladna) [inline playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerView.java]: why is this needed?

File: playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerWrapper.java [hunk 5]
```diff
@@ -1009,6 +1009,10 @@ public void onTracksInfoReady(PKTracks tracksReady) {
                     selectPreferredTracksLanguage();
                     preferredLanguageWasSelected = true;
                 }
+
+                if(tracks.getVideoTracks().size() == 0 && sourceConfig.getArtworkViewDrawable() != null && exoPlayerView != null) {
+                    exoPlayerView.setArtworkDrawable(sourceConfig.getArtworkViewDrawable());
```
Comment (giladna) [inline playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerWrapper.java:1025]: @GouravSna what is change media from audio only to video how view will be invisible?
do we need here elase to set to invisible?

File: playkit/src/main/java/com/kaltura/playkit/player/MediaPlayerView.java [hunk 6]
```diff
@@ -110,4 +110,9 @@ public void hideVideoSubtitles() {
     public void showVideoSubtitles() {
 
     }
+
+    @Override
+    public void artworkViewVisibility(boolean visibility) {
```
Comment (giladna) [inline playkit/src/main/java/com/kaltura/playkit/player/MediaPlayerView.java]: set prefix

File: playkit/src/main/java/com/kaltura/playkit/player/PKMediaSourceConfig.java [hunk 7]
```diff
@@ -42,20 +44,21 @@ public PKMediaSourceConfig(PKMediaSource source, PKMediaEntry.MediaEntryType med
...
+    PKMediaSourceConfig(PKMediaConfig mediaConfig, PKMediaSource source, PlayerSettings playerSettings, VRSettings vrSettings, Drawable artworkViewDrawable) {
         this.mediaSource = source;
         this.mediaEntryType = (mediaConfig != null && mediaConfig.getMediaEntry() != null) ? mediaConfig.getMediaEntry().getMediaType() : PKMediaEntry.MediaEntryType.Unknown;
         this.playerSettings = playerSettings;
         this.vrSettings = vrSettings;
         this.externalSubtitlesList = (mediaConfig != null && mediaConfig.getMediaEntry() != null && mediaConfig.getMediaEntry().getExternalSubtitleList() != null) ? mediaConfig.getMediaEntry().getExternalSubtitleList() : null;
+        this.artworkViewDrawable = artworkViewDrawable;
     }
 
     public PKMediaSourceConfig(PKMediaSource source, PKMediaEntry.MediaEntryType mediaEntryType, List<PKExternalSubtitle> externalSubtitlesList, PlayerSettings playerSettings) {
```
Comment (giladna) [inline playkit/src/main/java/com/kaltura/playkit/player/PKMediaSourceConfig.java:56]: no need here to support in drawable?

File: playkit/src/main/java/com/kaltura/playkit/player/PlayerController.java [hunk 8]
```diff
@@ -93,13 +93,16 @@ public void showVideoSurface() {
...
-
             }
 
             @Override
             public void showVideoSubtitles() {
                 setVideoSubtitlesVisibility(true);
+            }
 
+            @Override
+            public void artworkViewVisibility(boolean visibility) {
```
Comment (giladna) [inline playkit/src/main/java/com/kaltura/playkit/player/PlayerController.java]: set prefix

File: playkit/src/main/java/com/kaltura/playkit/player/PlayerView.java [hunk 9]
```diff
@@ -42,5 +43,10 @@ public PlayerView(Context context, AttributeSet attrs) {
 
     public abstract void showVideoSubtitles();
 
-    public void setSurfaceAspectRatioResizeMode(PKAspectRatioResizeMode resizeMode) {};
+    public void setSurfaceAspectRatioResizeMode(PKAspectRatioResizeMode resizeMode) {}
+
+    public void setArtworkDrawable(Drawable artworkDrawable){}
+
+    public abstract void artworkViewVisibility(boolean visibility);
```
Comment (giladna) [inline playkit/src/main/java/com/kaltura/playkit/player/PlayerView.java]: set

File: playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerView.java (hunk 3 above)
Comment (GouravSna) [reply playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerView.java]: Height and width; we are checking by getting the Intrinsic height and width. Basically If the drawable has no intrinsic height/width, such as a solid color, this method returns -1

File: playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerView.java (hunk 4 above)
Comment (GouravSna) [reply playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerView.java]: This is required for VRView, wrong placement here in ExoPlayerView. Needs to be shifted to VRView.

File: playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerWrapper.java (hunk 5 above)
Comment (GouravSna) [reply playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerWrapper.java:1025]: Fixed.

File: playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerView.java (hunk 4 above)
Comment (GouravSna) [reply playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerView.java]: This is required also for artwork view to change the aspect ratio for artworkview.

File: playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerView.java [hunk 10]
```diff
@@ -250,7 +261,26 @@ public void showVideoSubtitles() {
...
+
+    @Override
+    public void setArtworkViewVisibility(boolean isVisible) {
+        if (artworkDrawable != null) {
+            artworkView.setVisibility(isVisible ? VISIBLE : GONE);
+        } else {
+            artworkView.setVisibility(GONE);
+        }
+
+        if (shutterView != null && shutterView.getVisibility() == VISIBLE) {
```
Comment (giladna) [inline playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerView.java:273]: 1. should be called only if isVisible is true
2.should be moved to be the first thing done in the method

File: playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerWrapper.java [hunk 11]
```diff
@@ -1009,6 +1009,13 @@ public void onTracksInfoReady(PKTracks tracksReady) {
                     selectPreferredTracksLanguage();
                     preferredLanguageWasSelected = true;
                 }
+
+                if (sourceConfig.getArtworkViewDrawable() != null && tracks.getVideoTracks().size() == 0 && exoPlayerView != null) {
```
Comment (giladna) [inline playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerWrapper.java]: please check for isEmpty instead of == 0

**Review by noamtamim — CHANGES_REQUESTED** [review]
Please don't do it like that. Instead, allow the app to set the subtitles view from outside (`Player.setSubtitlesView(View)`). Then the app can put the artwork on top of the player and put the subtitles on top of that.

File: playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerView.java (hunk 10 above)
Comment (GouravSna) [reply playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerView.java:273]: This logic is required for shutterview to keep it hide if the artworkview is there. Other wise it is visible in bg/fg case.

**Comment by giladna** [comment]
giving more responsibility to app to do artwork so this code is not needed

**Comment by OrenMe** [comment]
@giladna is this branch obsolete now - please delete branch if so. thanks.
